File: backend/store/AuthStore.py
import os
import pickle
import time
from . import JsonStore

authFilePath = os.path.join(os.getcwd(), "data", "auth.data")


class AuthStore:
    def __init__(
        self,
        access_token: str | None = None,
        expires_at: str | None = None,
        refresh_token: str | None = None,
    ):
        self.access_token = access_token
        self.expires_at = expires_at
        self.refresh_token = refresh_token

    def existsDataFolder(self):
        return os.path.exists(self.getFolderPath())

    def getFolderPath(self):
        return os.path.dirname(authFilePath)

    def save(self):
        if self.existsDataFolder() == False:
            os.mkdir(self.getFolderPath())

        try:
            data = self

            print(data)

            JsonStore.saveFile(data.__dict__, authFilePath)
        except Exception as ex:
            print("Error saving auth data:", ex)

    def refresh(self, data):
        if data != None:
            self.access_token = data["access_token"]

            if data["refresh_token"] != None:
                self.refresh_token = data["refresh_token"]

            self.expires_at = time.time() + data["expires_in"]
        else:
            self.access_token = None
            self.refresh_token = None
            self.expires_at = None

        self.save()

    def loadFromRaw(self, raw):
        if raw == None:
            return

        self.access_token = raw["access_token"]
        self.refresh_token = raw["refresh_token"]
        self.expires_at = raw["expires_at"]

    def load(self):
        if self.existsDataFolder() == False:
            os.mkdir(self.getFolderPath())

        if not os.path.exists(authFilePath):
            self.save()
        else:
            try:
                raw = JsonStore.loadFile(authFilePath)

                if not raw == None:
                    self.loadFromRaw(raw)
            except Exception as ex:
                print("Error during unpickling object (Possibly unsupported):", ex)

File: api/src/class/AuthStore.ts
import fs from 'fs';
import path from 'path';

const authFilePath = path.join(process.cwd(), 'data', 'auth.data');

export default class AuthStore {
  access_token: string | null;
  expires_at: number | null;
  refresh_token: string | null;

  constructor(
    access_token: string | null = null,
    expires_at: number | null = null,
    refresh_token: string | null = null,
  ) {
    this.access_token = access_token;
    this.expires_at = expires_at;
    this.refresh_token = refresh_token;
  }

  existsDataFolder(): boolean {
    return fs.existsSync(this.getFolderPath());
  }

  getFolderPath(): string {
    return path.dirname(authFilePath);
  }

  save(): void {
    if (!this.existsDataFolder()) {
      fs.mkdirSync(this.getFolderPath(), {
        recursive: true
      });
    }

    try {
      const data = {
        access_token: this.access_token,
        expires_at: this.expires_at,
        refresh_token: this.refresh_token,
      };

      fs.writeFileSync(authFilePath, JSON.stringify(data));
    } catch (ex) {
      console.error('Error saving auth data:', ex);
    }
  }

  refresh(data: any): void {
    if (data !== null) {
      this.access_token = data.access_token;

      if ('refresh_token' in data) {
        this.refresh_token = data.refresh_token;
      }

      this.expires_at = (Date.now() / 1000 + data.expires_in).toString();
    } else {
      this.access_token = null;
      this.refresh_token = null;
      this.expires_at = null;
    }

    this.save();
  }

  loadFromRaw(raw: any): void {
    if (raw === null) {
      return;
    }

    this.access_token = raw.access_token;
    this.refresh_token = raw.refresh_token;
    this.expires_at = raw.expires_at;
  }

  load(): void {
    if (!this.existsDataFolder()) {
      fs.mkdirSync(this.getFolderPath(),
      {
        recursive: true
      });
    }

    if (!fs.existsSync(authFilePath)) {
      this.save();
    } else {
      try {
        const raw = fs.readFileSync(authFilePath, 'utf8');

        if (raw !== '') {
          const data = JSON.parse(raw);
          this.loadFromRaw(data);
        }
      } catch (ex) {
        console.error('Error during reading auth data:', ex);
      }
    }
  }
}
